### DISCUSSION TITLE
Codeium Chat functionality

### DISCUSSION THREAD

**Discussion by dmnsea** [veksha/cuda_codeium#11] [General]
I hope you will be able to change title if needed xD I just don't know if there's needed separate discussion especially about chat, or it will be enough to have general discussion at the moment.

I understand, that plugin is recently created and probably it don't has all functionality which you want to add to it.

There are 2 moments I've noticed yesterday, when staring server was fixed:

- chat context seems to be handled localy
- context become reset on cudatext restart
- each question and answer replaces content of the previous one without missing context (in terms of single session)

So current questions about chat:

- is it possible to manage context in any way? like manually reseting it, or save/restore. Or is it processed internally by codeium server and there's not found way to do it yet?
- is current behaviour with replacement is how you planned to make it, or it's just behave so for now and would be improved later? (at least working in appending mode instead of rewrite)

**Comment by veksha** [comment]
> I understand, that plugin is recently created and probably it don't has all functionality which you want to add to it.

i didn't plan to do all functionality, just simple autocompletion.  
some user suggested to add chat, it was interesting if i will be able to figure out how to accomplish this. it required new knowledge (protobuf library).

> * is it possible to manage context in any way? like manually reseting it, or save/restore. Or is it processed internally by codeium server and there's not found way to do it yet?

i think context must be managed by plugin. i monitored requests that was made from vscode extension to server, every single request to chat bot contains all chat context.

I'm used to not complicating unnecessarily.
you can always ask bot to `forget`, or restart CudaText :) restarting is very fast.

but if you want, you can try convince me that this simple plugin needs new commands for context management saving/restoring.
or send PR/patch.

> * is current behaviour with replacement is how you planned to make it, or it's just behave so for now and would be improved later? (at least working in appending mode instead of rewrite)

for me, personally, it's convenient like this. but suggestions are welcome.
It's interesting for me to try to add appending mode, (and maybe create option for this, enabled by default).

**Reply by dmnsea** [reply]
> you can always ask bot to `forget`

do you mean just simply write `forget` as chat input and send it?

> It's interesting for me to try to add appending mode, (and maybe create option for this, enabled by default).

Thanks, I would wait for it :D

> but if you want, you can try convince me that this simple plugin needs new commands for context management saving/restoring.
or send PR/patch.

it's interesting, but most likely i'll try somewhere in july. anyway any suggestions from which parts of code to start and etc?

**Reply by veksha** [reply]
> do you mean just simply write `forget` as chat input and send it?
 
yes.
```
User: forget
Bot: I have forgotten the previous conversation. How may I help you now?
```

> anyway any suggestions from which parts of code to start and etc?

`self.messages` contains context. you can make a plan. write pseudo-code, and send here. we can discuss.
how new workflow would look like with your patch.

**Reply by dmnsea** [reply]
well, in general, just thoughs currently, but I would like to add save option and use either json or pickle to dump context into file, and add restore option to load context from file.

also, if it will be possible to save context in json, then it should be relatively easy to edit if needed, so for quite long context it makes possible to revert some part of dialog

my target is just keep context between sessions and when switching between projects. restoring context from file would be much easier than starting almost same dialog from scratch to provide correct context for the actual questions.

i would think about plan, but let it be like "general steps" for a start:
1. add `save context` and `restore context` to pluging menu
2. figure out context format and write methods to process it
3. probably make kinda "pretty-print" on save if json will be used, so it would be easier to edit if needed
4. check the cudatext plugins api to understand how to make file opening and saving dialogs
5. bind them to the actual management methods

**Reply by veksha** [reply]
sounds ok 👍

**Reply by veksha** [reply]
@NEK-RA
added append_mode
